kernel/kernel-impl/src/main/java/org/sakaiproject/messaging/impl/UserMessagingServiceImpl.java

```diff
@@ -69,9 +82,35 @@ public class UserMessagingServiceImpl implements UserMessagingService {
     @Setter    private ResourceLoader resourceLoader;
 
     private ExecutorService executor;
+    private PushService pushService;
+    private String publicKey = "";
+    private boolean pushEnabled = false;
+
+    private final static ObjectMapper objectMapper = new ObjectMapper();
 
     public void init() {
+
+        Security.addProvider(new BouncyCastleProvider());
+
         executor = Executors.newFixedThreadPool(20);
+
+        this.pushEnabled = serverConfigurationService.getBoolean("portal.notifications.push.enabled", false);
+
+        if (pushEnabled) {
+            String home = serverConfigurationService.getSakaiHomePath();
+
+            String publicKeyFileName = serverConfigurationService.getString("portal.notifications.push.publickey", "sakai_push.key.pub");
+            String privateKeyFileName = serverConfigurationService.getString("portal.notifications.push.privatekey", "sakai_push.key");
+            try {
+                publicKey = String.join("", Files.readAllLines(Paths.get(home, publicKeyFileName)));
```

Comment:
Is providing a public/private key required? It looks like the library _can_ [run without it.](https://github.com/web-push-libs/webpush-java/wiki/VAPID#vapid). Could probably just put a info message or something in here if they're not provided but still go through with the setup?

Alternatively perhaps we could [generate vapid keys](https://stackoverflow.com/questions/44149608/generate-a-vapid-keys-in-java-and-pass-them-to-javascript-pushmanager) in the code with `KeyPaidGenerator `if not provided? Maybe this won't be a big deal but just less configuration is usually better.

Reply:
Dude, that is an awesome idea. I'll try it out now.

Reply:
In my testing, on Firefox, you have to have the VAPID keypair there. Mozilla's push service just rejected the push. I'll double check.